bpf/bpf_lxc.c

```diff
@@ -679,6 +679,30 @@ static __always_inline int handle_ipv4_from_lxc(struct __ctx_buff *ctx,
 		}
 	}
 
+#ifdef ENABLE_EGRESS_GATEWAY
+	{
+		struct egress_info *info;
+		struct endpoint_key key = {};
+
+		info = lookup_ip4_egress_endpoint(ip4->saddr, ip4->daddr);
+		if (!info)
+			goto skip_egress_gateway;
+
+		/* Encap and redirect the packet to egress gateway node through a tunnel.
+		 * Even if the tunnel endpoint is on the same host, follow the same data
+		 * path to be consistent. In future, it can be optimized by directly
+		 * direct to external interface.
+		 */
+		ret = encap_and_redirect_lxc(ctx, info->tunnel_endpoint, encrypt_key,
+					     &key, SECLABEL, monitor);
```

Comment:
Is the reason why the packet gets SNATed with the wrong IP by iptables? If I ran it with `--enable-bpf-masquerade` it doesn't have the issue. 

Actually for we are thinking of treating `egress_ip` as `tunnel_endpoint`, so we don't need to know which node has the `egress_ip`. For example, we can do this on the nodes: 
```
cilium bpf egress add 10.11.1.194 192.168.33.13/32 192.168.33.100 192.168.33.100
```

Reply:
> Is the reason why the packet gets SNATed with the wrong IP by iptables? If I ran it with --enable-bpf-masquerade it doesn't have the issue.

We had not enabled BPF masquerading, so that's probably it. We could likely add a clear dependency on BPF masquerading for the first version and fatal if users tries to enable egress gateway without it.

Reply:
sounds like a plan!

Reply:
Hey Paul, I found this doesn't have to depend on `--enable-bpf-masquerade`. I found that adding the following rule will fix the issue. I don't see any harm to disable conntrack for cilium_host's input path. What do you think. 

```
-A CILIUM_PRE_raw -i cilium_host -j NOTRACK
```

Reply:
Wouldn't that cause us to skip any reverse SNAT for traffic coming from the tunnel and going to the hostns? It feels a bit hackish and we should probably fix the packet path instead. We currently go `lxc -> cilium_vxlan -> cilium_host -> eth0` when we should just have `lxc -> eth0`.

Maybe we can require `enable-bpf-masquerade` for the first version and improve after the feature freeze?

Reply:
Confirmed that it is the case. This is probably the best approach now. I've applied the change already.